ovn-tester/ovn_utils.py

```diff
@@ -377,141 +518,128 @@ def acl_add(
 
     def lb_add_to_switches(self, lb, switches):
-        cmd = ' -- '.join([f'ls-lb-add {s} {lb.uuid}' for s in switches])
-        self.run(cmd=cmd)
+        with self.idl.transaction(check_error=True) as txn:
+            for s in switches:
+                txn.add(self.idl.ls_lb_add(s, lb.uuid))
 
     def lb_remove_from_routers(self, lb, routers):
-        cmd = ' -- '.join([f'lr-lb-del {r} {lb.uuid}' for r in routers])
-        self.run(cmd=cmd)
+        with self.idl.transaction(check_error=True) as txn:
+            for r in routers:
+                txn.add(self.idl.lr_lb_del(r, lb.uuid))
 
     def lb_remove_from_switches(self, lb, switches):
-        cmd = ' -- '.join([f'ls-lb-del {s} {lb.uuid}' for s in switches])
-        self.run(cmd=cmd)
-
-    def wait_until(self, cmd=""):
-        self.run("wait-until " + cmd)
+        with self.idl.transaction(check_error=True) as txn:
+            for s in switches:
+                txn.add(self.idl.ls_lb_del(s, lb.uuid))
 
     def sync(self, wait="hv", timeout=DEFAULT_CTL_TIMEOUT):
-        self.run(
-            f'--timeout={timeout} --wait={wait} sync', timeout=(timeout + 1)
-        )
+        with self.idl.transaction(
+            check_error=True, timeout=timeout, wait_type=wait
+        ):
+            pass
 
-    def start_daemon(self, nb_cluster_ips, enable_ssl):
-        if enable_ssl:
-            remote = ','.join([f'ssl:{ip}:6641' for ip in nb_cluster_ips])
-        else:
-            remote = ','.join([f'tcp:{ip}:6641' for ip in nb_cluster_ips])
-        # FIXME: hardcoded args, are these really an issue?
-        cmd = (
-            f'--detach --pidfile --log-file --db={remote} '
-            f'-p /opt/ovn/ovn-privkey.pem -c /opt/ovn/ovn-cert.pem '
-            f'-C /opt/ovn/pki/switchca/cacert.pem'
-        )
-        stdout = StringIO()
-        self.run(cmd=cmd, stdout=stdout)
-        self.socket = stdout.getvalue().rstrip()
 
-    def stop_daemon(self):
-        if len(self.socket):
-            self.sb.run(cmd=f'ovs-appctl -t {self.socket} exit')
+class SBIdl(sb_impl_idl.OvnSbApiIdlImpl, Backend):
+    def __init__(self, connection):
+        super(SBIdl, self).__init__(connection)
 
+    @property
+    def _connection(self):
+        # Shortcut to retrieve the lone Connection record. This is used by
+        # NBTransaction for synchronization purposes.
+        return next(iter(self.db_list_rows('Connection').execute()))
 
-class OvnSbctl:
-    def __init__(self, sb):
-        self.sb = sb
 
-    def run(self, cmd="", stdout=None, timeout=DEFAULT_CTL_TIMEOUT):
-        self.sb.run(
-            cmd="ovn-sbctl --no-leader-only " + cmd,
-            stdout=stdout,
-            timeout=timeout,
+class OvnSbctl:
+    def __init__(self, sb, connection_string, inactivity_probe):
+        i = connection.OvsdbIdl.from_server(
+            connection_string, "OVN_Southbound"
         )
+        c = connection.Connection(i, inactivity_probe)
+        self.idl = SBIdl(c)
```

Comment:
It looks like this will monitor the whole SbDB.  I don't think we want that.  For example, we don't want to receive all the logical flows and spend a lot of time and memory on parsing and storing them.  We should subclass the connection like neutron does and only register tables that we're interested in.  In this case we only need a `Connection` and `Chassis` tables.

Reply:
Something like this, I guess (some re-naming is required):
```
@@ -612,6 +637,16 @@ class OvnNbctl:
         ):
             pass
 
+from ovsdbapp.backend.ovs_idl import idlutils
+class BaseOvnSbIdl(connection.OvsdbIdl):
+    schema = "OVN_Southbound"
+
+    @classmethod
+    def from_server(cls, connection_string):
+        helper = idlutils.get_schema_helper(connection_string, cls.schema)
+        helper.register_table('Chassis')
+        helper.register_table('Connection')
+        return cls(connection_string, helper)
+
 
 class SBIdl(sb_impl_idl.OvnSbApiIdlImpl, Backend):
     def __init__(self, connection):
@@ -626,9 +660,7 @@ class SBIdl(sb_impl_idl.OvnSbApiIdlImpl, Backend):
 
 class OvnSbctl:
     def __init__(self, sb, connection_string, inactivity_probe):
-        i = connection.OvsdbIdl.from_server(
-            connection_string, "OVN_Southbound"
-        )
+        i = BaseOvnSbIdl.from_server(connection_string)
         c = connection.Connection(i, inactivity_probe)
         self.idl = SBIdl(c)
 
```

Reply:
I applied this fix. Thanks!